### PR TITLE
add endpoint `search`

### PR DESCRIPTION
**Pull request by remi-dupre**
This mostly reuses code from `instant_answer` but I tried to change the API a bit to be more consistent with what we usually use internally between erdapfel and Idunn:

 - the answer format is simpler, either consisting of the same format as `get_place` (through key `place`) or the format of `get_bbox_places` (through key `bbox_places`)
 - we don't usually raise error codes 404 when we have an empty geocoder result, I tried to stick with this by returning an empty result

The only slight behavior change is that there is no truncating of the input query, maybe we should add a global parameter for this? I've also thought about skipping normalization since it would seem weird to "ignore or change part of what is in the search bar" but this seems like a free potential improvement of results?

Part of the logs may be weird with the current state of the branch, I'm still thinking there may be some global pattern that could be built since fetching the context from the logger is a common issue we have (eg. how we build extra infos in `nlu_client` or in this case understanding what endpoint is currently being used).

### CODE REVIEW

File: idunn/api/search.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    # Direct geocoding query
+    bragi_response = await bragi_client.raw_autocomplete(
+        {"q": query.q, "lang": query.lang, "limit": 5}
+    )
+    geocodings = (feature["properties"]["geocoding"] for feature in bragi_response["features"])
+    geocodings = sorted(
+        (
+            (result_filter.rank_bragi_response(query.q, feature), feature)
+            for feature in geocodings
+            if result_filter.check_bragi_response(query.q, feature)
```
Comment (amatissart) [inline idunn/api/search.py:105]: As this search is explicitly about Maps, we should probably be more permissive than for Instant Answers. For example, some extra tokens (that do not match the result) should probably be permitted.   
But this could probably be discussed in a separate PR.

File: idunn/api/search.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+def search_single_place(place_id: str, lang: str):
+    try:
+        place = place_from_id(place_id, follow_redirect=True)
+    except Exception:
+        logger.warning("search: Failed to get place with id '%s'", place_id, exc_info=True)
+        return no_search_result()
+
+    detailed_place = place.load_place(lang=lang)
+    return IdunnAutocomplete(features=[detailed_place])
```
Comment (amatissart) [inline idunn/api/search.py:39]: this function is no longer used

File: idunn/api/instant_answer.py [hunk 3]
```diff
@@ -181,7 +181,7 @@ async def get_instant_answer_intention(intention, query: str, lang: str):
             maps_frame_url=maps_urls.get_places_url(intention.filter, no_ui=True),
         )
 
-    return build_response(result, query=query, lang=lang)
+    return result
```
[the remark this answers is not in the document]
Comment (amatissart) [reply idunn/api/instant_answer.py:184]: This change still breaks the instant answer in some cases. For example on `?q=restaurant`, this function will return a 204 `Response`, and `build_response()` will fail.
It would probably be clearer to annotate `get_instant_answer_single_place` and `get_instant_answer_intention` to show that the return type is always `Response`. 
Anyway, it's no longer related to this PR, so maybe this change should be reverted for now, and be handled separately.

File: idunn/api/search.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                "request": {
+                    "query": query,
+                    "lang": lang,
+                }
+            },
+        )
+    return IdunnAutocomplete()
+
+
+async def search(query: QueryParams = Depends(QueryParams)) -> IdunnAutocomplete:
```
Comment (amatissart) [inline idunn/api/search.py:30]: Only `q` and `lang` query are supported. So the documentation would be misleading for this route.  For example `?nlu=false` would not have any effect.

Is there any obstacle to handle the same query parameters as `/v1/autocomplete` ?

Comment (remi-dupre) [reply idunn/api/search.py]: I agree with you that it sounds nicer to support the same set of parameters as `/v1/autocomplete`, I had some issues with it on a previous implementation but there is nothing holding it back now, this is fixed!

File: idunn/api/urls.py [hunk 5]
```diff
@@ -74,6 +75,11 @@ def get_api_urls(settings):
             response_model=IdunnAutocomplete,
             response_model_exclude_unset=True,
         ),
+        APIRoute(
+            "/search",
+            search,
+            response_model=IdunnAutocomplete,
+        ),
```
Comment (amatissart) [inline idunn/api/urls.py:82]: For consistency with "/autocomplete", `methods` and `response_model_exclude_unset` should be defined too.

File: idunn/api/search.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+) -> IdunnAutocomplete:
+    """
+    Perform a query which is intended to display a relevant result directly (as
+    opposed to `autocomplete` which gives a list of plausible results).
+
+    Similarly to `instant_answer`, the result will need some quality checks.
+    """
+    query.q = normalize(query.q)
+
+    if query.lang in nlu_allowed_languages:
```
Comment (amatissart) [inline idunn/api/search.py:41]:
```suggestion
    if query.nlu and query.lang in nlu_allowed_languages:
```

Comment (remi-dupre) [reply idunn/api/search.py]: We also want to keep this option overridden by [AUTOCOMPLETE_NLU_SHADOW_ENABLED](https://github.com/Qwant/idunn/blob/master/idunn/utils/default_settings.yaml#L130), right?

Comment (amatissart) [reply idunn/api/search.py]: Well, this option has never been actually used, and is probably not very useful now that the `nlu=true` is used by the application in most cases.  
So I think it can be ignored for `/search`.

**Review by amatissart — APPROVED** [review]
